pex/pex_builder.py

```diff
@@ -675,6 +677,28 @@ def build(
             safe_delete(path)
         os.rename(tmp_pex, path)
 
+    def set_sh_boot_script(
+        self,
+        pex_name,  # type: str
+        targets,  # type: Targets
+        python_shebang,  # type: Optional[str]
+    ):
+        if not self._frozen:
```

Comment:
Erm. Hopefully this warning never fires in Pex code today; if so I'd rather it be an assert / raise. Pex is not intended for use as a library, just as a stable CLI API. Of course I'd like to circle back and cleanup many APIs internally any how, like making this builder undego state tranitions by returning new types that expose the newly valid methods only when valid, but that's a 10% benefit thing and doesn't have the pull currently.

Reply:
Will do. Fwiw I do use pex as a library and extend the pex builder. Generally it has been fine adapting to the fact that pex builder itself isn't guaranteed stable during updates.

Reply:
Aha - OK. You are warned at any rate. I definitely reserve the right to freely refactor all except the CLI.

Reply:
I would be interested in why the CLI isn't good enough though! If you want to relay some of that information via issues or offline channels, I'm definitely interested if its somethng more fundamental than legacy reasons.

Reply:
Sure, I would be happy to. Generally we use PEXBuilder to have fine grained control over the chroot path of source files. Adding whole directories isn't enough to handle all the subtleties of source roots for us.

Reply:
opened #1907